core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java

```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 }
                 instanceMetaDataService.save(instanceMetaDataEntry);
             }
+            primaryGWSelectionFallbackIfNecessary(primaryIgSelected, instanceGroups);
+
             return newInstances;
         } catch (CloudbreakImageNotFoundException | IllegalArgumentException ex) {
             throw new CloudbreakServiceException("Instance metadata collection failed", ex);
         }
     }
 
+    private boolean setupInstanceMetaDataType(boolean primaryIgSelected, Optional<InstanceMetaData> terminatedPrimaryGwWhichShouldBeRestored,
+            InstanceMetaData instanceMetaDataEntry, InstanceGroup ig) {
+        if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
+            if (ig != null) {
+                LOGGER.info("Instance group type: {}", ig.getInstanceGroupType());
+                if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
+                    primaryIgSelected = setupGatewayInstance(primaryIgSelected, terminatedPrimaryGwWhichShouldBeRestored, instanceMetaDataEntry);
```

Comment:
I don't think it's a good idea to change the method parameter value here. either introduce a new variable and assign the initial value to the parameter or reorganize the code a bit and use returns at multiple lines